src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,958 @@
+                            }
+                            // Retain non-empty batches
+                            !batch.is_empty()
+                        })
+                    }
+
+                    // 4.3. Reclock `source_frontier`
+                    //      If any FromTime in source frontier could possibly be reclocked to this
+                    //      binding then we must maintain our capability to emit data at that time
+                    //      and not compact past it.
+                    if source_frontier
+                        .frontier()
+                        .iter()
+                        .any(|t| !cur_binding.less_equal(t))
+                    {
+                        reclocked_source_frontier.insert(cur_time);
+                    }
+
+                    // Advance to the next binding if there is any
+                    match remap.peek() {
+                        Some((_, t, _)) => cur_time = t.clone(),
+                        None => break,
+                    }
+                }
+
+                // STEP 5. Downgrade capability and compact remap trace
+                capset.downgrade(&reclocked_source_frontier.borrow());
+                remap_since = reclocked_source_frontier;
+                for (_, t, _) in remap_trace.iter_mut() {
+                    t.advance_by(remap_since.borrow());
+                }
+                consolidation::consolidate_updates(&mut remap_trace);
+                remap_trace
+                    .sort_unstable_by(|(_, t1, _): &(_, IntoTime, _), (_, t2, _)| t1.cmp(t2));
+            }
+
+            // STEP 6. Tidy up deferred updates
+            //         Deferred updates are represented as a list chain batches where each batch
+            //         contains two times the updates of the batch proceeding it. This organization
+            //         leads to a logarithmic number of batches with respect to the outstanding
+            //         number of updates.
+            deferred_source_updates.sort_unstable_by_key(|b| Reverse(b.len()));
+            if !new_source_updates.is_empty() {
+                deferred_source_updates.push(new_source_updates);
+            }
+            loop {
+                match (deferred_source_updates.pop(), deferred_source_updates.pop()) {
+                    (Some(a), Some(b)) if a.len() >= b.len() / 2 => {
+                        deferred_source_updates.push(a.merge_with(b));
+                    }
+                    (a, b) => {
+                        deferred_source_updates.extend(b);
+                        deferred_source_updates.extend(a);
+                        break;
+                    }
+                }
+            }
+        }
+    });
+
+    (Box::new(pusher), reclocked.as_collection())
+}
+
+/// A batch of differential updates that vary over some partial order. This type maintains the data
+/// as a set of disjoint chains that allows for efficient extraction of batches given a frontier.
```

Comment:
I'm not sure what "disjoint" implies here. There can be overlap in the chains, like the same update could be in multiple chains. Perhaps just remove the word?